Improve clipboard format support in X11 client

ilammy opened the pull request:
Fixes in named formats (copy-paste HTML), raw format transfers (copy-paste anything between remote session), and some other things. See commit messages for more technical details.

Overview
-----------

### Named format support

As one may learn from MS-RDPECLIP spec, clipboard formats are identified either by _format ID_ (32-bit integer) or by _format name_ (arbitrary text string). Fixed-ID formats have fixed IDs assigned to them while named formats can have different IDs on different machines. The tricky part is in that cliprdr protocol is always using IDs to identify clipboard formats in protocol requests, so it is important to use _server-side_ IDs when requesting data from the server. XFreeRDP was doing it wrong, used client-side IDs instead, and expectedly was getting refusals from the server.

These commits revamp format identification and make sure that correct format IDs are used everywhere:

64d5e83 winpr/clipboard: add CF_RAW clipboard format ID
5b511a6 client/X11: improve clipboard format search functions
fed1907 client/X11: improve named clipboard format support

### Miscellaneous fixes

Some miscellaneous minor fixes:

36d3489 client/X11: correctly trim terminating null bytes from strings
f413325 client/X11: send clipboard format data errors correctly
eecd628 client/X11: remove unused function

### Raw transfer support

In the past XFreeRDP had a unique feature: ability to pass arbitrary clipboard formats between two XFreeRDP remote sessions, not limiting them to formats generally supported by X11 applications (basically, text, HTML, and some image formats). For example, this allowed to copy-paste formatted text between MS Office applications without losing any style information. Some parts of this feature still survived, but several other changes have broken it.

These commits restore the missing parts of the raw transfer mechanism:

842ea62 client/X11: refactor raw clipboard transfer indication
e4352bb client/X11: transfer raw clipboard format lists
81a584f client/X11: transfer raw clipboard format data

Generally, the idea stays the same:

  1. Session A receives a format list, assumes clipboard ownership, and advertises raw format list via a special mechanism.

  2. Session B is able to tell that an XFreeRDP session owns the clipboard, extracts raw format list and sends it to the remote side.

  3. Session B receives a format data request, and forwards the request to session A via a special mechanism.

  4. Session A passes the request to its remote side to get the actual data.

  5. The received data is passed intact from session A to session B, and then to the remote side of the session B.

The steps 1 and 5 were broken by previous changes and needed a fix.


Verification
------------

### Named formats

The only named format supported by XFreeRDP is `"HTML Format"`. It can be verified by copying anything from a remote browser into a local application supporting "text/html" (e.g, LibreOffice Writer). HTML should be kept intact: i.e., hyperlinks should be present, bold text should be pasted as bold, etc. Obviously, the same should be true in reverse.

Note that currently FreeRDP _does not remove_ `<!--StartFragment-->` and `<!--EndFragment-->` from HTML data, which results into them being shown as editorial comments by LibreOffice Writer. While additional metadata in comments is [a documented feature of `"HTML Format"`](https://msdn.microsoft.com/en-us/library/aa767917(v=vs.85).aspx), I am not sure that they should be removed by FreeRDP.

### Raw transfers

To check this case it is necessary to open two remote sessions and try transfering formats not supported by XFreeRDP. A good example would be `"Rich Text Format"`, a named format used by WordPad. The point is in that WordPad does not use `"HTML Format"` so it is not possible to confuse them. If raw transfer works correctly, rich formatting should be preserved when text is transferred between remote sessions.

To be absolutely sure that the feature works correctly one can build FreeRDP with enabled cliprdr dumps (`-DWITH_DEBUG_CLIPRDR=ON`) and observe communications over the _cliprdr_ channel.

### Affected known issues

This pull request fixes issues #1414, #2287, #2841.

It may fix #1433, but I am not able to verify it.

bmiklautz commented:
@freerdp-bot test

hardening commented:
@ilammy thank you for the contribution, could you rebase you PR on the master ? Thank you.

ilammy commented:
@hardening I have successfully rebased the changes onto current master, and retested them. Everything seems to be fine.

Resolved conflicts and changes in the original patch set:
- **client/X11: improve clipboard format search functions**
   + Resolved merge conflict around `clipboard->serverFormats` initialization in xf_cliprdr_server_format_list(). Removed the condition on `clipboard->numServerFormats` at the same place as it is always true now due to CF_RAW.
- **client/X11: refactor raw clipboard transfer indication**
   + @nfedera has  removed the old logic of xf_cliprdr_is_self_owned() and \_FREERDP_CLIPRDR_ID property altogether in a4580923e77b4690a5fe90a8e37a042bc5a8fa06. And he did it for a good reason: long time ago xf_cliprdr_is_self_owned() was used for raw transfers, but now it is used only to check whether _the current_ XFreeRDP window owns the clipboard. It should stay so. However, we do need a property to indicate raw transfer capability, so I have reintroduced it in the same way as it was implemented before, but it is now independent of xf_cliprdr_is_self_owned().
    + Updated commit message for this patch accordingly.
- **client/X11: transfer raw clipboard format lists**
   + Added more error logging with WLog_ERR as it is done in surrounding code by a4580923e77b4690a5fe90a8e37a042bc5a8fa06

akallabeth commented:
LGTM +1

bmiklautz commented:
@freerdp-bot test

hardening commented on client/X11/xf_cliprdr.c:
```diff
@@ -190,14 +271,151 @@ static UINT xf_cliprdr_send_data_response(xfClipboard* clipboard, BYTE* data, in
+	}
+
+	Stream_Read_UINT32(s, *numFormats);
+
```
nitpicking, but perhaps we could check that this is a reasonable value...

hardening commented on client/X11/xf_cliprdr.c:
```diff
@@ -190,14 +271,151 @@ static UINT xf_cliprdr_send_data_response(xfClipboard* clipboard, BYTE* data, in
+			goto error;
+		}
+
+		Stream_Read_UINT32(s, formats[i].formatId);
```
We should check that we don't go after the end of the stream.

akallabeth replied:
@hardening Isn't the check in line 352 sufficient?

hardening replied:
@akallabeth we have to check at each loop turn as the name has a variable length.

hardening commented:
Except my remark LGTM. 
@ilammy correct this and I will merge it.

akallabeth commented:
@ilammy Could you fix the two remarks from @hardening? They pervent a merge of your very nice pull ;)

ilammy commented:
@akallabeth Sorry, I got carried away by... other things :zzz: 

I have pushed a new commit that improves on the issues pointed out by @hardening:
* the number of formats to be read is now checked for sanity (255 should be enough for anybody)
* format names are now read more carefully in order to avoid reading past the buffer's end

hardening commented on client/X11/xf_cliprdr.c:
```diff
@@ -190,14 +273,170 @@ static UINT xf_cliprdr_send_data_response(xfClipboard* clipboard, BYTE* data, in
+		}
+
+		formats[i].formatName = strndup(formatName, formatNameLength);
+		Stream_Seek(s, formatNameLength + 1);
```
very nice handling of the \0 in the stream, congrats !
